FINAL /islam_eldsoke

islameldsoke opened the pull request:
- Apply MVP design pattern.
- Apply Rx with Retrofit.

islameldsoke committed:
First Commit

islameldsoke committed:
Get Tasks Commit

saraelmoghazy commented on app/src/main/res/layout/fragment_add_task.xml:
```diff
@@ -0,0 +1,26 @@
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:id="@+id/btn_addNewTask"
+        android:text="add"/>
```
plz add all strings in strings file

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/data/sourse/TasksRepository.java:
```diff
@@ -0,0 +1,87 @@
+
+    boolean cachedTasks = false;
+
+    List<Task> taskList = new ArrayList<>();
```
remove extra lines between class members , just one line after class name , one line before each method

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/task/TasksFragment.java:
```diff
@@ -0,0 +1,107 @@
+
+public class TasksFragment extends Fragment implements ITaskView {
+
+    private static final String ARG_PARAM1 = "param1";
```
remove unused paramaters

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/add_tasks/AddTaskFragment.java:
```diff
@@ -0,0 +1,84 @@
+
+        View view = inflater.inflate(R.layout.fragment_add_task, container, false);
+
+        title = view.findViewById(R.id.tv_title);
```
it's better to use butterknife

islameldsoke committed:
Second Commit

-Add new task funcatioality
-Add task detail fragment
-Apply previous reviews

islameldsoke committed:
Third Commit

-Adding some documintion
-Remove Spaces
-using lambada
-using butterknife

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/add_tasks/IAddTaskPresenter.java:
```diff
@@ -0,0 +1,13 @@
+ * @author islam
+ */
+
+public interface IAddTaskPresenter extends BasePresenter {
```
:+1:

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/add_tasks/IAddTaskView.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface IAddTaskView extends BaseView {
+    void setTitle(String title);
+    void setDescription(String description);
```
method not used

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/add_tasks/AddTaskFragment.java:
```diff
@@ -0,0 +1,87 @@
+    @BindView(R.id.btn_addNewTask)
+    Button addNewTask;
+    AddTaskPresenter addTaskPresenter;
+    public AddTaskFragment() {
```
no need for constructor

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/add_tasks/AddTaskFragment.java:
```diff
@@ -0,0 +1,87 @@
+
+        View view = inflater.inflate(R.layout.fragment_add_task, container, false);
+        ButterKnife.bind(this , view);
+        addTaskPresenter = new AddTaskPresenter(this,Injection.provideTasksRepository());
```
Injection is a great idea

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/add_tasks/AddTaskFragment.java:
```diff
@@ -0,0 +1,87 @@
+    public void onResume() {
+        super.onResume();
+
+        addNewTask.setOnClickListener(view -> addTaskPresenter.saveTask(new Task(UUID.randomUUID().toString()
```
use butterknife @onClick , if you will use setOnClickListener put it onCreate not onResume (it's define one time only).

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/add_tasks/AddTaskFragment.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    public void setTitle(String title) {
```
there is no need for setTitle ,setDescription in IAddTaskView , we want to make something when success (back to pervious fragment) , and failure(show error message in toast)

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/add_tasks/AddTaskFragment.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    public void setPresenter(Object presenter) {
```
why you are not using setPresenter ?

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/data/sourse/remote/ApiClient.java:
```diff
@@ -0,0 +1,21 @@
+
+    public static Retrofit getClient() {
+        if (retrofit == null)
+            retrofit = new Retrofit.Builder().baseUrl(BASE_URL).addConverterFactory(GsonConverterFactory.create()).build();
```
try not to exceed line length , make code statement in multiple lines

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/data/sourse/remote/RemoteTaskRepo.java:
```diff
@@ -0,0 +1,66 @@
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class RemoteTaskRepo implements TasksDataSource {
```
it's better to have repo for each module (save/get/show details) , we want to have 1 view --> 1 presenter --> 1 repo

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/task/ITaskPresenter.java:
```diff
@@ -0,0 +1,11 @@
+ */
+
+public interface ITaskPresenter extends BasePresenter {
+    void loadTasks();
```
there is no need for loadTasks  , you already call loadtasks in start method , use it private in presenter

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/taskdetail/TaskDetailFragment.java:
```diff
@@ -0,0 +1,82 @@
+    Task task;
+
+
+    public TaskDetailFragment() {
```
no need for constructor in fragment.

saraelmoghazy commented on app/src/main/java/com/ntg/user/mvpsample/taskdetail/TaskDetailFragment.java:
```diff
@@ -0,0 +1,82 @@
+                             Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_task_detail, container, false);
+        ButterKnife.bind(this , view);
+
```
add presenter to decide what to do when TaskDetailFragment start

islameldsoke committed:
new changes

-seperate the repo
-convert details fragment to mmet MVP

islameldsoke committed:
Test Commit

Add TestPresenter

islameldsoke committed:
Rx opertor

islameldsoke committed:
SubTask UI

islameldsoke committed:
Rx Progress

islameldsoke committed:
Add new functionality

-Add Validation

islameldsoke committed:
[WIP] optimizing

-Add callback for adding tasks and subtasks.
-Add loading.
-Add sum tests.
-Add subtask list UI.

islameldsoke committed:
RxError Handel

islameldsoke committed:
Rx custom

islameldsoke committed:
Add error handling + retrofit adaption

islameldsoke committed:
Final Custom RxAdapter

islameldsoke committed:
Dagger 2

islameldsoke committed:
apply dagger

islameldsoke committed:
Replace click item listner with Rx
